scheduler/src/cook/kubernetes/compute_cluster.clj

```diff
@@ -363,14 +170,76 @@
+    - If config-file is specified, initializes the api file from the file at config-file
+    - If base-path is specified, sets the cluster base path
+    - If verifying-ssl is specified, sets verifying ssl
+    - If google-credentials is specified, loads the credentials from the file at google-credentials and generates
+      a bearer token for authenticating with kubernetes
+    - bearer-token-refresh-seconds: interval to refresh the bearer token"
+  [^String config-file base-path ^String google-credentials bearer-token-refresh-seconds verifying-ssl]
+  (let [api-client (if (some? config-file)
+                     (Config/fromConfig config-file)
+                     (ApiClient.))]
+    (when base-path
+      (.setBasePath api-client base-path))
+    (when (some? verifying-ssl)
+      (.setVerifyingSsl api-client verifying-ssl))
+    (when google-credentials
+      (with-open [file-stream (FileInputStream. (File. google-credentials))]
+        (let [credentials (GoogleCredentials/fromStream file-stream)
+              scoped-credentials (.createScoped credentials ["https://www.googleapis.com/auth/cloud-platform"
+                                                             "https://www.googleapis.com/auth/userinfo.email"])
+              bearer-token (get-bearer-token scoped-credentials)]
+          (.scheduleAtFixedRate bearer-token-executor
+                                (make-bearer-token-refresh-task api-client scoped-credentials)
+                                bearer-token-refresh-seconds
+                                bearer-token-refresh-seconds
+                                TimeUnit/SECONDS)
+          (.setApiKey api-client bearer-token))))
+
+    api-client))
+
 (defn factory-fn
-  [{:keys [compute-cluster-name ^String config-file]} {:keys [exit-code-syncer-state
-                                                              trigger-chans]}]
+  [{:keys [compute-cluster-name
+           ^String config-file
+           base-path
+           google-credentials
+           verifying-ssl
+           bearer-token-refresh-seconds]
+    :or {bearer-token-refresh-seconds 300}}
+   {:keys [exit-code-syncer-state
+           trigger-chans]}]
   (let [conn cook.datomic/conn
         cluster-entity-id (get-or-create-cluster-entity-id conn compute-cluster-name)
-        api-client (Config/fromConfig config-file)
+        api-client (make-api-client config-file base-path google-credentials bearer-token-refresh-seconds verifying-ssl)
         compute-cluster (->KubernetesComputeCluster api-client compute-cluster-name cluster-entity-id
                                                     (:match-trigger-chan trigger-chans)
-                                                    exit-code-syncer-state (atom {}) (atom {}))]
+                                                    exit-code-syncer-state (atom {}) (atom {})
+                                                    (atom {:type :expected-state-map})
```

Comment:
Comment that these initial keys are only to make it easier to trace provenance when debugging.

Reply:
👍